File: problems/3871-count-commas-in-range-ii/solution.ex
defmodule Solution do
  @spec count_commas(n :: integer) :: integer
  def count_commas(n) when n < 1000, do: 0

  def count_commas(n) do
    count_commas(n, log1000(n), 0)
  end

  defp count_commas(_n, 0, acc), do: acc

  defp count_commas(n, tier, acc) do
    tier_start = pow1000(tier)
    tier_end = min(n, pow1000(tier + 1) - 1)
    count_commas(n, tier - 1, acc + tier * (tier_end - tier_start + 1))
  end

  defp pow1000(0), do: 1
  defp pow1000(k), do: 1000 * pow1000(k - 1)

  def log1000(n) when n < 1000, do: 0
  def log1000(n), do: 1 + log1000(div(n, 1000))
end

File: problems/3871-count-commas-in-range-ii/solution.erl
-spec count_commas(N :: integer()) -> integer().
count_commas(N) when N < 1000 ->
    0;
count_commas(N) ->
    count_commas(N, floor_log1000(N), 0).

count_commas(_N, 0, Acc) ->
    Acc;
count_commas(N, Tier, Acc) ->
    TierStart = pow_1000(Tier),
    TierEnd = min(N, pow_1000(Tier + 1) - 1),
    count_commas(N, Tier - 1, Acc + Tier * (TierEnd - TierStart + 1)).


pow_1000(0) -> 1;
pow_1000(K) -> 1000 * pow_1000(K - 1).

floor_log1000(N) when N < 1000 -> 0;
floor_log1000(N) ->
    1 + floor_log1000(N div 1000).
